### ISSUE TITLE
Safely handle non-DAG cases when using TopologicalOrderIterator

### ISSUE THREAD

**Issue by kigh-ota** [jgrapht/jgrapht#1095]
```
 * JGraphT version: 1.5.1
 * Java version (java -version)/platform:  11
```

**Issue**

Hello,

In version 1.1.0 (https://github.com/jgrapht/jgrapht/pull/374), `TopologicalOrderIterator` was added DAG validation and it throws a generic `IllegalArgumentException` when it detects that the graph is cyclic.
What I would like to do is to `TopologicalOrderIterator` with possibly cyclic graphs and handle a non-DAG case as an error.
I can handle the non-DAG case by catching `IllegalArgumentException` and seeing whether the message string is equal to "Graph is not a DAG".
However this is not very safe, because the message string is not public and may change as an implmentation detail.

An alternative way is to check beforehand that the graph is acyclic using `CycleDetector`.
But this adds slight computational cost for DAG cases.

Can `TopologicalOrderIterator` throw a more specific exception (say `NotDirectedAcyclicGraphException`)?
It enables to handle a non-DAG case easily just by catching the specific exception.
One drawback is that this change would break existing codes that expect `IllegalArgumentException` to be thrown.

Could this approach be acceptable? If so I am willing to create a pull request.

Thanks.

**Steps to reproduce (small coding example)**

**Expected behaviour**

**Other information**

**Comment by jsichi** [comment]
We could make the new exception extend IllegalArgumentException.  @d-michail waddaya think?

**Comment by d-michail** [comment]
Yes, a `NotDirectedAcyclicGraphException` extending `IllegalArgumentException` sounds like a good idea.

**Comment by kigh-ota** [comment]
Thank you. Can I submit a PR?

**Comment by jsichi** [comment]
Yes, please do!

**Comment by jsichi** [comment]
Released in 1.5.3
